File: web/js/renderers/sphereRenderer.js
// Generic 3D sphere rendering with lighting, labels, and color helpers.

// Lighten a color by amount (0-1). color is [r,g,b,a].
export function lighten(color, amount) {
  return [
    Math.min(255, color[0] + (255 - color[0]) * amount),
    Math.min(255, color[1] + (255 - color[1]) * amount),
    Math.min(255, color[2] + (255 - color[2]) * amount),
    color[3] !== undefined ? color[3] : 255,
  ];
}

// Darken a color by amount (0-1). color is [r,g,b,a].
export function darken(color, amount) {
  return [
    color[0] * (1 - amount),
    color[1] * (1 - amount),
    color[2] * (1 - amount),
    color[3] !== undefined ? color[3] : 255,
  ];
}

// Blend two colors. t=0 → a, t=1 → b. Colors are [r,g,b] or [r,g,b,a].
export function blendColor(a, b, t) {
  t = Math.max(0, Math.min(1, t));
  return [
    a[0] + (b[0] - a[0]) * t,
    a[1] + (b[1] - a[1]) * t,
    a[2] + (b[2] - a[2]) * t,
    (a[3] ?? 255) + ((b[3] ?? 255) - (a[3] ?? 255)) * t,
  ];
}

function rgba(c) {
  return `rgba(${Math.round(c[0])},${Math.round(c[1])},${Math.round(c[2])},${(c[3] !== undefined ? c[3] : 255) / 255})`;
}

// Draws a 3D-shaded sphere with a light direction.
export function drawSphere(ctx, cx, cy, radius, baseColor, lightDirX, lightDirY) {
  const highlightX = cx + lightDirX * radius * 0.35;
  const highlightY = cy + lightDirY * radius * 0.35;

  const lightColor = lighten(baseColor, 0.6);
  const darkColor = darken(baseColor, 0.7);
  const darkerColor = darken(baseColor, 0.9);

  const grad = ctx.createRadialGradient(
    highlightX, highlightY, 0,
    highlightX, highlightY, radius * 1.8
  );
  grad.addColorStop(0, rgba(lightColor));
  grad.addColorStop(0.35, rgba(baseColor));
  grad.addColorStop(0.75, rgba(darkColor));
  grad.addColorStop(1, rgba(darkerColor));

  ctx.fillStyle = grad;
  ctx.beginPath();
  ctx.arc(cx, cy, radius, 0, Math.PI * 2);
  ctx.fill();
}

// Draws a text label with shadow.
export function drawLabel(ctx, text, x, y, fontSize, color, align = 'center') {
  ctx.font = `${fontSize}px sans-serif`;
  ctx.textAlign = align;
  ctx.textBaseline = 'middle';

  // Shadow
  ctx.fillStyle = `rgba(0,0,0,0.59)`;
  ctx.fillText(text, x + 1, y + 1);

  // Text
  ctx.fillStyle = rgba(color);
  ctx.fillText(text, x, y);
}

// Draws a dashed orbit path (ellipse).
export function drawOrbitPath(ctx, cx, cy, radiusX, radiusY) {
  ctx.save();
  ctx.strokeStyle = 'rgba(100,100,150,0.31)';
  ctx.lineWidth = 1;
  ctx.setLineDash([6, 4]);
  ctx.beginPath();
  ctx.ellipse(cx, cy, radiusX, radiusY, 0, 0, Math.PI * 2);
  ctx.stroke();
  ctx.setLineDash([]);
  ctx.restore();
}

File: MoonSimulation/Renderers/SphereRenderer.cs
using SkiaSharp;

namespace MoonSimulation.Renderers;

/// <summary>
/// Generic 3D sphere rendering with lighting, labels, and color helpers.
/// </summary>
public static class SphereRenderer
{
    /// <summary>
    /// Draws a 3D-shaded sphere with a light direction.
    /// </summary>
    public static void DrawSphere(SKCanvas canvas, SKPoint center, float radius,
        SKColor baseColor, SKPoint lightDir, float ambientLight = 0.15f)
    {
        using var paint = new SKPaint { IsAntialias = true };

        // Offset highlight toward the light source
        float highlightX = center.X + lightDir.X * radius * 0.35f;
        float highlightY = center.Y + lightDir.Y * radius * 0.35f;

        // Radial gradient from highlight to dark edge
        var lightColor = Lighten(baseColor, 0.6f);
        var darkColor = Darken(baseColor, 0.7f);

        using var shader = SKShader.CreateRadialGradient(
            new SKPoint(highlightX, highlightY),
            radius * 1.8f,
            new SKColor[] { lightColor, baseColor, darkColor, Darken(baseColor, 0.9f) },
            new float[] { 0f, 0.35f, 0.75f, 1f },
            SKShaderTileMode.Clamp);

        paint.Shader = shader;
        canvas.DrawCircle(center, radius, paint);
    }

    /// <summary>
    /// Draws a text label with shadow.
    /// </summary>
    public static void DrawLabel(SKCanvas canvas, string text, SKPoint position, float fontSize,
        SKColor color, SKTextAlign align = SKTextAlign.Center)
    {
        using var font = new SKFont(SKTypeface.Default, fontSize);
        using var shadowPaint = new SKPaint
        {
            IsAntialias = true,
            Color = new SKColor(0, 0, 0, 150)
        };
        using var textPaint = new SKPaint
        {
            IsAntialias = true,
            Color = color
        };

        canvas.DrawText(text, position.X + 1, position.Y + 1, align, font, shadowPaint);
        canvas.DrawText(text, position.X, position.Y, align, font, textPaint);
    }

    /// <summary>
    /// Draws a dashed orbit path (ellipse).
    /// </summary>
    public static void DrawOrbitPath(SKCanvas canvas, SKPoint center, float radiusX, float radiusY)
    {
        using var pathPaint = new SKPaint
        {
            IsAntialias = true,
            Style = SKPaintStyle.Stroke,
            StrokeWidth = 1f,
            Color = new SKColor(100, 100, 150, 80),
            PathEffect = SKPathEffect.CreateDash(new float[] { 6, 4 }, 0)
        };

        canvas.DrawOval(center, new SKSize(radiusX, radiusY), pathPaint);
    }

    public static SKColor Lighten(SKColor color, float amount)
    {
        return new SKColor(
            (byte)Math.Min(255, color.Red + (255 - color.Red) * amount),
            (byte)Math.Min(255, color.Green + (255 - color.Green) * amount),
            (byte)Math.Min(255, color.Blue + (255 - color.Blue) * amount),
            color.Alpha);
    }

    public static SKColor Darken(SKColor color, float amount)
    {
        return new SKColor(
            (byte)(color.Red * (1 - amount)),
            (byte)(color.Green * (1 - amount)),
            (byte)(color.Blue * (1 - amount)),
            color.Alpha);
    }

    public static SKColor BlendColor(SKColor a, SKColor b, float t)
    {
        t = Math.Clamp(t, 0, 1);
        return new SKColor(
            (byte)(a.Red + (b.Red - a.Red) * t),
            (byte)(a.Green + (b.Green - a.Green) * t),
            (byte)(a.Blue + (b.Blue - a.Blue) * t),
            (byte)(a.Alpha + (b.Alpha - a.Alpha) * t));
    }
}
